### ISSUE TITLE
Grunt won't fail

### ISSUE THREAD

**Issue by michaeltorbert** [semperfiwebdesign/all-in-one-seo-pack#2454]
Grunt is autocorrecting (locally) and so doesn't fail when it comes across problems. I think it's this: https://github.com/semperfiwebdesign/all-in-one-seo-pack/blob/master/Gruntfile.js#L48

We need to decide whether we want that behavior or not.

**Comment by arnaudbroes** [comment]
@EkoJR will have to back me up here but I think we could simply change the order of our Grunt tasks and run PHPCS before PHPCBF so that Travis will fail when there are errors that need to be fixed, but so we can also still run PHPCBF as a separate task in PHPStorm to fix the issues.

**Comment by EkoJR** [comment]
@arnaudbroes Sounds possible, but running local tests phpcbf is intended to run prior, and I wonder if putting it in last will turn out confusing; getting errors to errors that no longer exists.

I'm leaning more towards adding a seperate test, so that there would be two types. 1) `default` used by travis 2) `dev` for local development tests. 

`default` would be strictly error checking and code sniffing.

`dev` would contain extra resources considered helpful, but intrusive to travis; like phpcbf is.

**Comment by michaeltorbert** [comment]
> @arnaudbroes Sounds possible, but running local tests phpcbf is intended to run prior, and I wonder if putting it in last will turn out confusing; getting errors to errors that no longer exists.
> 
> I'm leaning more towards adding a seperate test, so that there would be two types. 1) `default` used by travis 2) `dev` for local development tests.
> 
> `default` would be strictly error checking and code sniffing.
> 
> `dev` would contain extra resources considered helpful, but intrusive to travis; like phpcbf is.

That sounds like a great idea.

**Comment by contactashish13** [comment]
why do we not like grunt auto-correcting the errors?

**Comment by arnaudbroes** [comment]
@contactashish13 we don't want Grunt auto-correcting errors in Travis so that build will fail if there still are errors that haven't been resolved. I believe @EkoJR has set up a second batch of Grunt tasks for dev use which still includes PHPCBF.

**Comment by contactashish13** [comment]
@arnaudbroes phpcbf corrects errors that it can auto correct. That's what we depend on locally as well as developers. For example, it would be a terrible waste of time to manually fix each line that should have spaces instead of tabs and phpcbf does that for you. That is after all the task of a JS task runner. As far as errors that phpcbf cannot fix, if I commit them grunt will fail because they have to be fixed manually. 

I'm afraid I still do not understand the intent of this issue.

**Comment by michaeltorbert** [comment]
Ben can correct me if I'm wrong, but I think the idea is that PHPCBF is correcting errors in travisci (locally only within the travis container), so then it never fails coding standards checks.

Autocorrecting locally on a dev machine is good of course.

I think the plan is to figure out a way autocorrect can happen during dev but not in travis builds.

> why do we not like grunt auto-correcting the errors?

**Comment by contactashish13** [comment]
@michaeltorbert that's what I've mentioned that that's what its job is - to correct errors that can be corrected. It cannot and does not correct errors that need manual intervention. Are we talking about simple phpcbf-correctable errors or manual phpcbf-non-correctable errors? Both run identically on local and travis.

**Comment by michaeltorbert** [comment]
This is in the context of travisci, not a developer's local machine. 
My understanding is that the errors corrected by grunt are only temporary and local to travis, which causes subsequent checks to pass when they should fail, if grunt wasn't previously run by a developer with the changes committed.

**Comment by contactashish13** [comment]
Ok so do you want jobs to fail because of a (simple) error that WordPress was spelled incorrectly (just as an illustration) or had tabs instead of spaces and the developer knew that phpcbf would run and correct these things?

**Comment by EkoJR** [comment]
From what I've found, `phpcs` is built to work in conjunction with `phpcbf`. Even with `phpcbf` removed from install, I still end up with an error log saying at the end...

> PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY.

...and I haven't found a way to get around requiring phpcbf. It also appears the php_codesniffer devs don't have an alternative. https://github.com/squizlabs/PHP_CodeSniffer/issues/981

I also found an error with the latest node_js and wpcs installed. Doesn't occur on older versions.

`Warning: PHP Parse error:  syntax error, unexpected '?', expecting variable (T_VARIABLE) in tmp/wpcs/bin/class-ruleset-test.php on line 56`

I think there needs to be a way to apply the grunt changes with a patch. I saw there was some mention of the concept. I'm not a huge fan of automated changes, but it looks like it is the expected behavior to be doing.

**Comment by michaeltorbert** [comment]
https://github.com/marketplace/codefactor looks like it can "automatically" do the changes for php codesniffer.

**Comment by EkoJR** [comment]
I took a look into applying a patch through travis, and it does seem possible. However, it does require a secure api key to be setup for each repository, which is then added to travis encrypted variable, then added to the travis yml

This is one of the examples...
https://gist.github.com/willprice/e07efd73fb7f13f917ea

The problem with this, it requires...

1) Seems limited to 1 repo; which won't work with PRs coming from a different repo.
2) Multiple tokens would have to be used in order to make a commit on a given Repo/PR.
3) Access tokens are encrypted on a public file. Recent git news (repos held for ransom) makes this seem like a Risk factor.
4) There is also a potential infinite build loop https://github.com/travis-ci/travis-ci/issues/9329

I went back to seeing if I can get grunt to fail, but no luck.

It may be possible to use the command line instead of grunt, and force a fail if the code sniffing execution fails.
